### ISSUE TITLE
Decorator repo naming convention?

### ISSUE THREAD

**Issue by Anthropic** [Textalk/angular-schema-form#393]
We currently have add-ons prefixed with **angular-schema-form-** should there be a convention for decorators that can be picked up so I can start looking at a **Material Design** based decorator whenever I find free time?

**Comment by davidlgj** [comment]
Hi @Anthropic,

good question! `angular-schema-form-decorator` is a bit long as a prefix :) Hmmm... do you have any suggestions? For visibility on the schemaform.io we'll do a bower tag, so that can be that as long as we want. 

And btw I'm all for merging in decorators into the main repo, especially something useful like a material design based one. So just a fork could be the simplest way to get going. 

I've started (twice) on a angular-material based decorator, I'll see if I can find the latest and get it into shape enough to push it into a branch.

**Comment by Anthropic** [comment]
I started one before they had drop downs without noticing the futility of such an effort.

Can we add a param to the bower file to indicate and then just name like an add on? angular-schema-form-materialdesign would be an appropriate repo name.

I was actually thinking it could be good to split out even the currently included one and have it as a dependency instead so it could act as a seed repo. But I'd be OK with bundling both if you can't for see too many being added.

**Comment by davidlgj** [comment]
I'm on the fence :) Bundling is a bit simpler, and easier to install for people, on the other hand we already have ObjectPath as a dependecy.... 

Maybe it would be cleaner with a repo per decorator. We could try do to some sort of CDN (such as cdnjs) bundling for easy installation.

**Comment by Dervisevic** [comment]
Bundling with addons and decorators would be awesome!

**Comment by davidlgj** [comment]
@Dervisevic CDN bundling or all in main repo?

**Comment by Dervisevic** [comment]
CDN style custom installs of ASF

**Comment by davidlgj** [comment]
:+1: Yeah that would be awesome

**Comment by Anthropic** [comment]
I would love a builder like what jQ UI have if that's what you mean?

**Comment by davidlgj** [comment]
Yes, exactly! raven-js has something similar as well

**Comment by davidlgj** [comment]
Ok @Anthropic I think you changed my mind :) One repo per decorator and when the first decorator is ready we'll move the bootstrap decorator to it's own repo, and update shcemaform.io (and the readme) with a new list of decorators. Does that sound good?

How about `angular-schema-form-decorator` as a bower tag for decorator repos?

**Comment by Anthropic** [comment]
Is there substantial dev involved for that though?

**Comment by Anthropic** [comment]
Yeah sounds good to me :)

**Comment by davidlgj** [comment]
Don't know if it is.... we'll have to look into how other projects solved it.

**Comment by davidlgj** [comment]
@Anthropic I can get a repo up with my latest stab at a decorator this afternoon and maybe we could collaborate a bit?

**Comment by Anthropic** [comment]
Sounds good to me,  that's my last message today, about to walk into evening session of Mad Max :)

**Comment by davidlgj** [comment]
@Anthropic I created a repo :fireworks: 

https://github.com/Textalk/angular-schema-form-material

It's very very rough, but you can clone it and check out the example.

Highlights: 
  *  Only input and textarea (but bootstrap files for the other to ease implementation)
  *  No real solution where description should go
  *  I had problem adding `class="{{form.fieldHtmlClass}}"` to the input, angular-material doesn't like that.
  *  The default theme, or any other theme, in angular material doesn't kick in so it doesn't look that good.

I haven't found the exact problem with the "theme" of angular material disappearing, it seems like it has something to do with the fact that we use `$compile` in the link function of the decorator directive, we do this since its not until then we know the type of the field hence the template. fyi this manual compile in the link phase is also the biggest performance hogger. 

So I've started work on a "builder" service that would be a big change but could probably fix the problems with material  and the performance issues. It would also enable us to drop the `<bootstrap-decorator>` tags around every field making it a bit more flexible with the layout. I'm trying to do it in a fashion that is backwards compatible though so it doesn't break current usage. The plan is that instead of directives like tabs, fieldset, array having an ng-repeat with directives we build up the entire tree of templates from the start and thereby need no `$compile` in the link function. Hopefully I'm able to make some headway this week with it.

**Comment by Anthropic** [comment]
@davidlgj that sounds great, much cleaner, I was actually wondering how hard it would be to add access to the top level form scope to templates too, but I guess that's another ticket...

**Comment by davidlgj** [comment]
Good evening @Anthropic, the first version of the "builder" is in this branch: https://github.com/Textalk/angular-schema-form/tree/feature/builder  

I made it backwards compatible so not to break peoples CSS, this means the ordinary bootstrap decorator basically doesn't use it yet. But I thought the material decorator could start using it. So here is the matching branch for that: https://github.com/Textalk/angular-schema-form-material/tree/feature/with-builder

And I got themes working! (probably in the older non-builder version as well) by setting `alwaysWatchTheme` to `true` (`$mdThemingProvider.alwaysWatchTheme(true);`). Still no real idea *why* it fails though.

**Comment by Anthropic** [comment]
:clap:
Good afternoon @davidlgj I'll take a look, sounds very promising!

**Comment by davidlgj** [comment]
God morning @Anthropic,  small change. No tests broke and since it doesn't change bootstrap decorator I merged `feature/builder` into `development` branch.

I also merged the `feature/with-builder` into `develop` branch of the material repo, and fixed bower.json, it should be easier to test now.
